models/User.go

```diff
@@ -1677,15 +1793,32 @@ func PerformOrderFillTransaction(ask *Ask, bid *Bid, stockTradePrice uint64, sto
 		bid.IsClosed = oldBidIsClosed
 
 		tx.Rollback()
-		return AskUndone, BidUndone, nil
+	}
+
+	if cashLeft < MINIMUM_CASH_LIMIT {
+		l.Errorf("Check1: Failed. Not enough cash. Returning reserved cash back to user.")
+		if err := bid.Close(tx); err != nil {
+			errorHelper("Unable to close bid. Rolling back. Error: %+v", err)
+			return AskUndone, BidUndone, nil
+		}
+		SaveBidCancelOrderTransaction(bid, biddingUser, tx)
+		if err := tx.Commit(); err != nil {
+			errorHelper("Error while commiting. Rolling back. Error: %+v", err)
+			return AskUndone, BidUndone, nil
+		}
+		go updateDataStreams(nil, nil, nil, nil)
+		go SendNotification(biddingUser.Id, fmt.Sprintf("Your Buy order#%d has been closed due to insufficient cash", bid.Id), false)
+		return AskUndone, BidDone, nil
 	}
 
 	// Tax Calculation for asking user
 	transactionSummary, askTaxTransaction := getTaxForAskingUser(tx, ask.StockId, stockTradeQty, stockTradePrice, askingUser)
 
 	// Update transaction summary table for asking user
 	if err := tx.Save(transactionSummary).Error; err != nil {
-		return errorHelper("Error updating the transaction summary for asking user. Rolling back. Error : +%v", err)
+		errorHelper("Error updating the transaction summary for asking user. Rolling back. Error : +%v", err)
+		return AskUndone, BidDone, nil
```

Comment:
Why is BidDone? If we're rolling back the transaction, nothing is done, right? In which case returning that from errorHelper will be enough I think.

Reply:
Why have you changed the errorHelper definition? It used to return BidUndone, and now you're returning BidDone everywhere?

Reply:
Could you perhaps update errorHelper definition to take an AskOrderFillStatus and a BidOrderFillStatus as arguments and return them? Instead of calling the error function, and then returning what you want to return separately

Reply:
@thakkarparth007 Yeah my bad.
@gauthamk97 Returning directly from errorHelper gives less control over what can be returned. Could always pass as arguments but didn't wanna overload the function with too many paramters.

Reply:
@PaiAkshay998 Pass it as arguments instead. It is much cleaner compared to how you've currently implemented it

Reply:
Anyway now it's always gonna return XUndone, so you can return from helper directly no?

Reply:
Yeah it was like that but except for one particular case. And it did make sense to use errorHandler there as well because errorHelper does a a couple of other stuff as well apart from a simple rollback. So if that had to use errorHelper, in that case as well, we'll also have to return differently.

Reply:
Which case? I'm surprised there's a case where there's an error but one's done and one's not (especially after doing initial checks)

Reply:
@thakkarparth007 If on attempting to fill an order (PerformOrderFillTransaction), it's found that the bidder does not have enough cash, we should close the Bid but leave the Ask untouched. In that case, we return AskUndone,BidDone